(RK-308) Allow respecting `forge` declaration in puppetfile

Magisus opened the pull request:
This commit adds a new setting `forge.allow_puppetfile_override`, that
when true will cause a `forge` declaration in a Puppetfile to override
the `forge.baseurl` setting for contacting the Forge. This allows users
to configure per-environment Forge URLs, for example, if an internal
Forge is used for a test environment and the live Forge for production.
Previously, a user would have had to choose one or the other. The
setting is false by default, since many users have erroneous `forge`
declarations in their Puppetfiles that were previously completely
ignored.

Magisus committed:
(RK-308) Allow respecting `forge` declaration in puppetfile

This commit adds a new setting `forge.allow_puppetfile_override`, that
when true will cause a `forge` declaration in a Puppetfile to override
the `forge.baseurl` setting for contacting the Forge. This allows users
to configure per-environment Forge URLs, for example, if an internal
Forge is used for a test environment and the live Forge for production.
Previously, a user would have had to choose one or the other. The
setting is false by default, since many users have erroneous `forge`
declarations in their Puppetfiles that were previously completely
ignored.

Magisus commented on lib/r10k/module_loader/puppetfile.rb:
```diff
@@ -114,6 +115,13 @@ def add_module_metadata(name, info)
+          # This method will append a trailing slash to the string, in place, so we
+          # dup it to avoid different behavior in the value of our instance var
+          # when `allow_puppetfile_forge` is true vs false
+          PuppetForge.host = forge.dup
```
The dup is probably unnecessary, seeing as we don't really use this instance var for anything that I can see. Lemme know what you think. Also, if we're not using `@forge`... do we need the instance variable at all?

mwaggett replied:
yeah, this is confusing to me.. I would be in favor of removing `@forge` and the various `set_forge` methods if they're not actually doing anything, since it sorta seems like they're not? but it's unclear. Or perhaps replace the `set_forge` methods with ones that actually do something.. I would have imagined something calling `set_forge(forge_from_puppetfile)` and then having `PuppetForge.host = @forge` somewhere where it's actually being used? But I don't really know, everything is a mess haha.

Magisus replied:
As the comment at the top of this block says, `set_forge` is part of the Puppetfile parsing DSL. It is [called](https://github.com/puppetlabs/r10k/blob/4a832b7e0090ca2ec04c60f48ce43b98a6b16d5f/lib/r10k/module_loader/puppetfile/dsl.rb#L29) when the Puppetfile is [evaluated](https://github.com/puppetlabs/r10k/blob/4a832b7e0090ca2ec04c60f48ce43b98a6b16d5f/lib/r10k/module_loader/puppetfile.rb#L63-L64) by that DSL. Until now, it has set this instance variable, but been a noop since, so far as I could tell, the var isn't used (we were deliberately ignoring `forge` declarations in the Puppetfile).

The actual way to make the `forge` declaration be used is to set the `host` field for the `PuppetForge` gem. That's what we do when [reading it from settings](https://github.com/puppetlabs/r10k/blob/4a832b7e0090ca2ec04c60f48ce43b98a6b16d5f/lib/r10k/initializers.rb#L67). I didn't delete the instance variable, in case something else is expecting it be set, but I think maybe that's not the case. This seems even more likely given there's no getter for it. I can try that and see what happens.

mwaggett replied:
yeah, I know the DSL has a method that calls this, but I didn't see anywhere calling the DSL's `forge` method, so wondering if that could be removed too.

Magisus replied:
The DSL represents an implementation of how to evalute a Puppetfile as a Ruby script. So if a Puppetfile has
```
forge "my.awesome.forge

mod "puppetlabs/apache"
```
`forge` and `mod` there are Ruby function calls, where the definition is what's in the DSL class. So when `instance_eval` happens, we evaluate the Ruby script that is the Puppetfile in the context of the DSL class, which has those methods defined, effectively calling `forge`, then `mod`.

I was trying to look for a good example of using `instance_eval` this way, but it seems like most people use it to add methods to individual objects, instead of actually executing code directly like we're doing here. The docs might help though: https://www.rubydoc.info/stdlib/core/BasicObject:instance_eval

mwaggett replied:
oh right, ugh. didn't realize just `grep`ing wouldn't be enough 😅  but that makes sense.

mwaggett commented on CHANGELOG.mkd:
```diff
@@ -4,6 +4,7 @@ CHANGELOG
 Unreleased
 ----------
 
+- (RK-308) Provide a `forge.allow_puppetfile_override` setting, which when true causes a `forge` declaration in the Puppetfile to override `forge.baseurl` [#1214](https://github.com/puppetlabs/r10k/pull/1214)
```
(nit)
```suggestion
- (RK-308) Provide a `forge.allow_puppetfile_override` setting that, when true, causes a `forge` declaration in the Puppetfile to override `forge.baseurl`. [#1214](https://github.com/puppetlabs/r10k/pull/1214)
```

mwaggett commented on doc/puppetfile.mkd:
```diff
@@ -51,10 +51,9 @@ handles modules.
-R10k supports setting the Forge to use _globally_ in `r10k.yaml`. see [Configuration](/doc/dynamic-environments/configuration.mkd#baseurl) for details.
+from. This declaration is only respected if [`forge.allow_puppetfile_override`](/dynamic-environments/configuration.mkd#allow_puppetfile_override)
+is set to true in the main `r10k.yaml`. Otherwise, use [`forge.baseurl`](/doc/dynamic-environments/configuration.mkd#baseurl)
+to configure where modules should be downloaded from.
```
(nit)
```suggestion
to globally configure where modules should be downloaded from.
```

mwaggett commented on spec/unit/module_loader/puppetfile_spec.rb:
```diff
@@ -276,6 +276,30 @@ def expect_wrapped_error(error, pf_path, error_type)
+        expect(PuppetForge).not_to receive(:host=).with("my.custom.forge.com")
+        puppetfile.load!
+        # instance variable is set, but PuppetForge.host is not updated
+        expect(puppetfile.instance_variable_get(:@forge)).to eq("my.custom.forge.com")
```
Regardless of what we decide to do with `@forge`, I think it's kinda weird to test this, since we don't actually want to be using "my.custom.forge.com" in this case.. It would be ideal imo if `@forge` was equal to the default. ¯\_(ツ)_/¯

Magisus replied:
Yeah I see what you're saying, but I think I'm still kind of in favor of dropping the instance variable.

mwaggett replied:
that works too.

mwaggett commented on spec/unit/module_loader/puppetfile_spec.rb:
```diff
@@ -276,6 +276,30 @@ def expect_wrapped_error(error, pf_path, error_type)
+        puppetfile.load!
+        # instance variable is set, but PuppetForge.host is not updated
+        expect(puppetfile.instance_variable_get(:@forge)).to eq("my.custom.forge.com")
+        expect(PuppetForge.host).to eq("/")
```
Why doesn't this equal the default Forge value? It seems super weird to me to have various places where we set `PuppetForge.host` and that those places are different from where that host is actually used??

Magisus replied:
This is an artifact of the test: this is a unit test of the `ModuleLoader::Puppetfile`, so the thing that would normally set `PuppetForge.host` (settings init) doesn't happen here. If you notice above, I've provided a default value in a `before` block. If I didn't do that, the state bled between tests, because `host` is a class variable of `PuppetForge`. I could use a _different_ default value, but that doesn't change the nature of the test.

mwaggett replied:
oh right, I forgot you were resetting it by the time I got to this part haha. so if it was a single test where you didn't have the `before` block, `PuppetForge.host` would be the default value? Is that our default or the forge gem default?

Magisus replied:
It would be the Forge gem's default, because even with one test, our code that sets it is not getting executed in this context. I don't know if that default is something valid, or just `nil`, I'd have to check.

Magisus committed:
(maint) Remove unused `forge` instance var from ModuleLoader::Puppetfile

This class does not make any use of its `forge` instance variable, nor
does it expose it via a getter. Since it is unused, this commit removes
it and the associated tests.

Note that the old `Puppetfile` class still exposes a `forge` instance
variable that is hard-coded to a default, and does not respect either
the `forge.baseurl` setting or a `forge` declaration in a Puppetfile.
This var is also unused within r10k (`PuppetForge.host` is set directly
elsewhere), but might be used by people using the `Puppetfile` class by
itself.

mwaggett commented on CHANGELOG.mkd:
```diff
@@ -4,6 +4,7 @@ CHANGELOG
 Unreleased
 ----------
 
+- (RK-308) Provide a `forge.allow_puppetfile_override` setting that, when true causes a `forge` declaration in the Puppetfile to override `forge.baseurl`. [#1214](https://github.com/puppetlabs/r10k/pull/1214)
```
```suggestion
- (RK-308) Provide a `forge.allow_puppetfile_override` setting that, when true, causes a `forge` declaration in the Puppetfile to override `forge.baseurl`. [#1214](https://github.com/puppetlabs/r10k/pull/1214)
```

Magisus replied:
Honestly I don't really think either of these commas is necessary. The clause is short enough that it doesn't really need to be offset. But meh 🤷‍♀️

mwaggett replied:
well one was definitely wrong 😝  I usually just add them if it helps me parse the sentence ¯\_(ツ)_/¯

Magisus replied:
Yeah agreed, that was a typo. I put it in.

mwaggett commented on spec/unit/module_loader/puppetfile_spec.rb:
```diff
@@ -276,6 +267,27 @@ def expect_wrapped_error(error, pf_path, error_type)
+      it 'is ignored if `allow_puppetfile_override` is false' do
+        @path = File.join(PROJECT_ROOT, 'spec', 'fixtures', 'unit', 'puppetfile', 'forge-override')
+        puppetfile = R10K::ModuleLoader::Puppetfile.new(basedir: @path, overrides: { forge: { allow_puppetfile_override: false } })
+        expect(PuppetForge).not_to receive(:host=).with("my.custom.forge.com")
```
Is this necessary if we're testing the value of `PuppetForge.host` below?

Magisus replied:
What I like about this one is it ties us back to the specific value from the Puppetfile that we want to be ignored.

mwaggett commented on lib/r10k/module_loader/puppetfile.rb:
```diff
@@ -113,7 +111,10 @@ def add_module_metadata(name, info)
       def set_forge(forge)
-        @forge = forge
+        if @allow_puppetfile_forge
+          logger.debug _("Using Forge from Puppetfile: %{forge}") % { forge: forge }
```
Should we also log that we're ignoring the `forge` setting from the Puppetfile if `@allow_puppetfile_forge` isn't true?

Magisus replied:
I could go either way, it's default behavior and has been that way a really long time. But I can add it.

mwaggett replied:
Do we think we'd want to make this the default behavior in the next major release? If so, I think it could be good to log a message about it. ¯\_(ツ)_/¯ 
Otherwise, if we think this will be opt-in forever, I'm good with leaving it out.

Magisus replied:
I seriously doubt we will ever make this the default behavior. We're leaning away from having people put more things in their puppetfiles.

mwaggett replied:
In that case, I'm good with leaving out the logging. (sorry for the back and forth 😅 )

Magisus replied:
Having tested this out, I think I'll leave it in, actually. There's isn't any other logging currently at `debug` or `debug1` that tells you which forge you're using.
